src/Language/PureScript/Roles.hs

```diff
@@ -0,0 +1,97 @@
+  -- parameterising.
+  deriving (Show, Eq)
+
+-- |
+-- Given an environment and the qualified name of a type constructor in that
+-- environment, returns a list of (formal parameter name, role) pairs, in the
+-- order they are defined in the type definition.
+inferRoles :: Environment -> Qualified (ProperName 'TypeName) -> [(Text, Role)]
+inferRoles env tyName
+  | Just roles <- lookup tyName primRoles =
+      roles
+  | Just (_, DataType tvs ctors) <- M.lookup tyName envTypes =
+      -- For each constructor the type has, walk its list of field types and
+      -- accumulate a list of (formal parameter name, role) pairs. Then, walk
+      -- the list of defined parameters, ensuring both that every parameter
+      -- appears (with a default role of phantom) and that they appear in the
+      -- right order.
+      let ctorRoles = foldMap (foldMap walk . snd) ctors
+      in  map (\(tv, _) -> (tv, fromMaybe Phantom (lookup tv ctorRoles))) tvs
+  | otherwise
+      = []
+  where
+    envTypes = types env
+    -- This function is named walk to match the specification given in the "Role
+    -- inference" section of the paper "Safe Zero-cost Coercions for Haskell".
+    walk (TypeVar v) =
+      -- A type variable standing alone (e.g. `a` in `data D a b = D a`) is
+      -- representational.
+      [(v, Representational)]
+    walk (ForAll _ t _) =
+      -- We can walk under universal quantifiers. Note that this might appear
+      -- erroneous -- given a definition like `data D a = D (forall r. r -> a)`
+      -- we expect we'll produce `[("r", Representational)]` as part of the
+      -- recursive call's result. This is true, but fine, since this tuple will
+      -- never be looked up when building the final result, which only looks at
+      -- the variables defined as parameters to the type.
+      walk t
+    walk t
+      | Just (t1, t2s) <- splitTypeApp t =
+          case t1 of
+            -- If the type is an application of a type constructor to some
+            -- arguments, recursively infer the roles of the type constructor's
+            -- arguments. For each (role, argument) pair, recurse if the
+            -- argument is representational (since its use of our parameters is
+            -- important) and terminate if the argument is phantom.
+            TypeConstructor t1Name ->
+              let t1Roles = inferRoles env t1Name
+                  k (_v, role) ti = case role of
+                    Representational ->
+                      walk ti
+                    Phantom ->
+                      []
+              in  concat (zipWith k t1Roles t2s)
+            -- If the type is an application of any other type-level term, walk
+            -- the first term to determine what role contributions it makes
+            -- (e.g. in the case `data F a b = F (a (F a b))`, the use of `a` as
+            -- a function/constructor would trigger this case)
+            _ ->
+              walk t1
```

Comment:
It looks like this case should mark all  `t2s` as `representational`:

```
data F a b = F (a b)
```

As `a` is polymorphic, we don't know what its role annotation is going to be, therefore we should be careful and assume that its argument is `representational` (because it very well could be!). 

That means that for `F`, `b` would be inferred `representational` as well.
I _think_ the current implementation will infer `phantom` for `b`, allowing to coerce
`F Maybe Int` and `F Maybe String`

Reply:
> It looks like this case should mark all `t2s` as `representational`:

Or rather, just `walk` all `t2s`, as that will mark all type variables `representational` anyway, and ensure that we don't accidentally mark phantom things representational (which would be safe, but undesirable).

So:
```
data G a b = G (a (Phantom b))

data Phantom a = Phantom
```
will not get tricked into marking `b` as representational.

-----------------------

Note: this case is actually interesting, because it differs from what GHC does. Not having to consider `nominal` roles results in a non-trivial difference:

For `G` above, GHC would infer
```
type role G representational nominal
```
This is because with `nominal` as a possibility, we must assume that `a`'s argument can be `nominal`, therefore `Phantom b` is in a `nominal` role. However, `nominal` is "infectious": if `Phantom b` is nominal, then all type variable therein must be nominal too (`b` here). This being the case, GHC rightly marks `b` as nominal, and stops. In the paper, this is accounted for by the rules that mark all free variables nominal.

However, with only `phantom` and `representational` roles, we can do better than simply downgrading `nominal` to `representational`! `Phantom b` being in a `representational role` doesn't "infect" `b`: walking is sufficient to determine the above result.

As a result, GHC would never allow to `coerce` between `G Maybe Int` and `G Maybe String` - but we can!